chaosazure/machine/actions.py

```diff
@@ -144,10 +164,11 @@ def stress_cpu(filter: str = None,
         "Start stress_cpu: configuration='{}', filter='{}'".format(
             configuration, filter))
 
-    choice = __fetch_machine_at_random(filter, configuration, secrets)
+    choices = __fetch_machines_at_random(filter, 1, configuration, secrets)
+    choice = choices[0]
```

Comment:
We can get an exception here. 

In addition to that I would also follow the same strategy for this action as for other actions in this class. That means that we can pass "count" and stress cpu on more then on machine at once

Reply:
I fully agree. I was thinking about integrating *count* also into this function. What made me worry is the blocking property of the stress_cpu function. When we introduce the *count* for several machines here I'm wondering about i) how long those operations for e.g. five machines will take and ii) if it makes sense for the user to have the stress_cpu in a sequential way.

I would suggest to make the very useful stress_cpu function as a non-blocking function. What do you guys think?

Reply:
@Lawouach : I know that we once discussed a similar topic - running activities asynchronously. I know that there is the *background* option for activities. To summarize it here: when me, as a user, I want to stress the CPU on several machines (more than one) I probably want to do it asynchronously. With the *background* option I would need to describe the activity several times. In the opposite having the *count* parameter in the activity itself, the activity could run *count* asynch jobs itself to stress the CPU of the *count* different machines.

What do you think? Are there already some other use cases that demand or implement this?

Reply:
@buderre  @Lawouach 
My two coins 

>  I want to stress the CPU on several machines (more than one) I probably want to do it asynchronously

I would completely agree. What i want to do is to set stress CPU for 10 minutes on "x" machines and in the same time check steady state hypothesis.

> With the background option I would need to describe the activity several times

I would say not necessarily as we can design it that way that stress_cpu, restart_machine and other actions applies to more than one machine.  
That means than any action can operate on "x" resources and action is always synchronous from the perspective of the plugin developer. When developing new action it would be developer responsibility to make sure that you triggers stress CPU on "x" machines and then waits for "x" calls to finish. So inside it's kind of async but outside it is sync (if it make sense :D) as you always wait for all calls to finish. Then framework above would be responsible for making those actions calls (stress_cpu, restart) sync or async based on the selected attribute in chaos experiment.

> In the opposite having the count parameter in the activity itself, the activity could run count asynch jobs itself to stress the CPU of the count different machines
As we discussed before probably we can just use action filter to do it instead of additional count param. But for me this approach sounds more clean...

Reply:
Hey folks,

I think this discussion is sound and I like the approach you describe @mkaszub 
It does make sense that you offer a count to the activity and let the activity deal with how to implement it.

But to be fair, if you passed a count and set background to true, then the whole action would be in parrallel whether you do it for one or many VMs. However, you can also add a specific flag in your activity argument that let the experiment operator make a decision if this should be async or blocking.

Either way works for me in this case.

Reply:
> But to be fair, if you passed a count and set background to true, then the whole action would be in parrallel whether you do it for one or many VMs. However, you can also add a specific flag in your activity argument that let the experiment operator make a decision if this should be async or blocking.

This is valid point. 

So we have actions that can be sync/async (background switch) and in addition to that if actions applies to multiple resources we can:
 - perform action in parallel on all of them
 - perform one after another
In my opinion in both cases we should return if action(s) were success that means wait for call to finish. If we don't care about result and want the whole func to be async we should just set switch background to true.

That brings me to another question do we have any way of synchronization. This is what i want to achieve (let me know if it make sense)

stress_cpu (background: True)
probe
wait (action: stress_cpu - > check results)
shutdown_machine(background: False)
probe
turn_on_machine(background: False)
finito

Reply:
@mkaszub : When I see correctly you will find the answer for your question in the source code https://github.com/chaostoolkit/chaostoolkit-lib/blob/master/chaoslib/experiment.py#L270 The function waits for all background activities and synchronizes them.